lib/backend/pgbk/pgbk.go

```diff
@@ -430,56 +431,114 @@ func (b *Backend) Get(ctx context.Context, key backend.Key) (*backend.Item, erro
-				" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
-				" ORDER BY kv.key LIMIT $3",
-			nonNilKey(startKey), nonNilKey(endKey), limit,
-		).Query(func(rows pgx.Rows) error {
-			var err error
-			items, err = pgx.CollectRows(rows, func(row pgx.CollectableRow) (backend.Item, error) {
-				var key backend.Key
-				var value []byte
-				var expires time.Time
-				var revision revision
-				if err := row.Scan(&key, &value, (*zeronull.Timestamptz)(&expires), &revision); err != nil {
-					return backend.Item{}, err
+	const (
+		queryAsc = "SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv" +
+			" WHERE kv.key BETWEEN $1 AND $2 AND ($3::bytea is NULL or kv.key > $3) AND (kv.expires IS NULL OR kv.expires > now())" +
+			" ORDER BY kv.key ASC LIMIT $4"
+
+		queryDesc = "SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv" +
+			" WHERE kv.key BETWEEN $1 AND $2 AND ($3::bytea is NULL or kv.key < $3) AND (kv.expires IS NULL OR kv.expires > now())" +
+			" ORDER BY kv.key DESC LIMIT $4"
+
+		defaultPageSize = 1000
+	)
+	return func(yield func(backend.Item, error) bool) {
+		var exclusiveStartKey []byte
+		query := queryAsc
+		if params.Descending {
+			query = queryDesc
+		}
+
+		var totalCount int
+		for {
+			pageLimit := min(limit-totalCount, defaultPageSize)
+
+			items, err := pgcommon.RetryIdempotent(ctx, b.log, func() ([]backend.Item, error) {
+				batch := new(pgx.Batch)
+				// batches run in an implicit transaction
+				batch.Queue("SET transaction_read_only TO on")
+				// TODO(espadolini): figure out if we want transaction_deferred enabled
+				var items []backend.Item
+				batch.Queue(query, nonNilKey(params.StartKey), nonNilKey(params.EndKey), exclusiveStartKey, pageLimit).Query(func(rows pgx.Rows) error {
+					var err error
+					items, err = pgx.CollectRows(rows, func(row pgx.CollectableRow) (backend.Item, error) {
+						var key backend.Key
+						var value []byte
+						var expires time.Time
+						var revision revision
+						if err := row.Scan(&key, &value, (*zeronull.Timestamptz)(&expires), &revision); err != nil {
+							return backend.Item{}, err
+						}
+						return backend.Item{
+							Key:      key,
+							Value:    value,
+							Expires:  expires.UTC(),
+							Revision: revisionToString(revision),
+						}, nil
+					})
+					return trace.Wrap(err)
+				})
+
+				if err := b.pool.SendBatch(ctx, batch).Close(); err != nil {
+					return nil, trace.Wrap(err)
 				}
-				return backend.Item{
-					Key:      key,
-					Value:    value,
-					Expires:  expires.UTC(),
-					Revision: revisionToString(revision),
-				}, nil
+
+				return items, nil
 			})
-			return trace.Wrap(err)
-		})
+			if err != nil {
+				yield(backend.Item{}, trace.Wrap(err))
+				return
+			}
 
-		if err := b.pool.SendBatch(ctx, batch).Close(); err != nil {
+			if len(items) == pageLimit {
+				exclusiveStartKey = []byte(items[len(items)-1].Key.String())
```

Comment:
More defensive programming

```suggestion
			if len(items) >= pageLimit {
				items = items[:pageLimit]
				exclusiveStartKey = []byte(items[len(items)-1].Key.String())
```